### DISCUSSION TITLE
Improve the CP layout for very wide screens

### DISCUSSION THREAD

**Discussion by thupsi** [craftcms/cms#13968] [Ideas]
Craft's CP provides a very pleasant experience and I appreciate the constant improvements in the area of accessibility, author experience etc. 

As a designer myself, I'm often delighted by the little touches that are introduced in newer versions.

In any complex interface though, there are many things to take into account and IMO one area that could be improved would be the layout in larger screens.

### The problem
For users with very large/wide screens the CP content area becomes too wide, posing usability and readability problems: 
- Index tables become hard to read, with rows stretching from one side of the screen to the other
- Field layouts (when editing an entry) appear disorganized and hard to parse, with fields becoming too wide and drifting apart from each other
- Action buttons located in the corners require too much mouse movement 
- The lines of text in various fields, especially rich text fields, become too wide for comfortable reading.

### Proposed solution
In many interfaces (like here, in Github, for example) this problem is solved by imposing a maximum width on the main page content. I think Craft's CP would benefit from such an approach by:
- Having a fixed maximum width for the main content by default
- Or making this opt-in: Add a toggle in the user settings, or even a field to set the width on a per user basis.

### Basic Example

![image](https://github.com/craftcms/cms/assets/19857599/70fb17ab-7d17-4991-8545-836d84402fc3)

While not the worst example, the above image illustrates a scenario where the main content is too wide on a 4K screen.

![image](https://github.com/craftcms/cms/assets/19857599/91710e47-e81b-4518-bd93-6b85fcc4cbfc)

With some custom CSS the interface becomes much more manageable.

**Comment by aaronbushnell** [comment]
I think even a left-aligned version where a `max-width` being enforced doesn't look half bad! Though maybe the avatar in the top right needs aligned as well.

![Screenshot 2023-11-29 at 11 42 00 AM](https://github.com/craftcms/cms/assets/315202/ce1bb904-0eca-459b-a926-7e4d3fdea4de)

**Reply by thupsi** [reply]
Not bad for sure, I have also tested this! 😊

Though it can potentialy feel a bit too assymetrical, especially when the screen is ultrawide. I have no problem with the empty space, or the assymetry per se, just the fact that you may keep turning your head left to focus on the content.

**Comment by alexanderloewe** [comment]
I don't think everyone will like this. 
For my part, I think it's good to have more space in an admin area. Especially if you have very complex table formats or other elements that take up a lot of space.

**Reply by thupsi** [reply]
Each project is different, that's why I'm suggesting that this should be optional and customisable. 

Having said that, the only time I can imagine anyone using this much horizontal space meaningfully is with tables that have many columns, so essentialy in index pages only. Or in the dashboard with many widgets. In all other cases the fields just start floating around and become disjointed. (Though, in Craft 5 we will have indexes _inside_ entries, so that would potentialy benefit from more space.)

Another point we should take into acccount is that the UI should be fully usable in all possible resolutions, so it wouldn't be reasonable to take a large horizontal axis for granted. It would be more efficient to design things for a more constrained width as a starting point.

Anyway, food for thought as they say 🙂

**Reply by alexanderloewe** [reply]
You can do something like this with:
https://plugins.craftcms.com/cp-css

If you want to include such customisations in the Craft Core, I would be more in favour of incorporating the functionality to independently load a CSS file via the Core setting. This way, anyone can simply add the CSS themselves if they wish.

**Reply by thupsi** [reply]
That's how I do it right now, I have some standard customisations that are universally beneficial, and I apply in most of my projects. So, I'm thinking of making a plugin to automate such things.  I'll let you know how that goes!

**Reply by alexanderloewe** [reply]
Yes please.

**Comment by thupsi** [comment]
In light of a [recent PR](https://github.com/craftcms/cms/pull/14281), and the discussion that was followed by the partial reversal of said PR, I would like to make some additional observations:

### Resizing the window doesn't work.

One could think that an author or dev with a wide screen could simply resize their window to their prefered width, or maybe zoom in their browser. There are some problems though:

- If I resize my window or zoom in, I loose the space when I need it: I can't see all the columns of a wide table, the slideouts are too narrow, the slideouts cover too much of my content, I may always need to collapse the sidebar, etc

- If I choose to just zoom in, the UI does have better proportions, but - because the actual content width remains large on a large screen - I need to sit further from my screen to comfortably fit the content in my field of view.

### Variable content width could be the solution. 

Table views can demand _more horizontal space_. On the other hand _too much horizontal space_ is detrimental to the main content editing UI. Throw in possible idiosyncrasies of any given project or field layout, user preferences etc, and it becomes obvious that there simply is no one-size-fits-all solution: _The layout needs vary according to the task at hand_.

The good news is that we don't have to reinvent the wheel: There are examples of variable content widths for different views in admin UIs. A nice one is from Shopify:

**Table view:**
![shopify_table](https://github.com/craftcms/cms/assets/19857599/3c780f2d-cd1b-4570-a39c-6974cb97d081)


**Product edit screen:**
![shopify_form](https://github.com/craftcms/cms/assets/19857599/230f4d44-7e0c-46ee-b334-920d37e65c92)

### Even variable width has a problem.

Craft provides is a very flexible UI for managing content, and [Craft 5](https://craftcms.com/blog/craft-5-beta-released) will take this to the next level with nested Matrix fields and other content modelling features. This opens the possibility of complex nested layouts, large tables and _entire element indexes_ right inside an entry. So a fixed narrow width for the entry view could prove limiting under certain circumstances.

### What should we do then? 

I don't have THE perfect answer, but I can think of several options:

1. **Make the width configurable on the project level,** with a single global width for all views. This would not cover all bases, but at least users with strong preferences or specific project needs would not feel left-out.
2. **Have some views use the full width, and others use a constrained width (preferably configurable)**. This would be more flexible and would  give the opportunity for some sensible defaults.
3. **Have a width toggle** (full-width/narrow-width), maybe in the global header, that appears only on wide viewports. Even more flexible, but requires manual action from the user.
4. **A combination of 2 and 3**. This would be the most flexible, but also more complex, solution.

I'm not dogmatic on any of the above. Feedback and opinions are welcome! 🙂

**Reply by aelvan** [reply]
My vote is on 2), adapt the design so that it let's us have both wide element indexes, and reasonably wide edit pages.

**Reply by brandonkelly** [reply]
Yeah that’s what I’m leaning on as well; we just decided it’s best to hold off on introducing max width for the beta, until we’ve had time to think it through a bit more.
